referral.go

```diff
@@ -0,0 +1,193 @@
+// specifying a context that can interrupt the request.
+func (c *Client) UpdateReferralCustomerEmailWithContext(ctx context.Context, userId string, email string) (out *ReferralCustomer, err error) {
+	req := referralCustomerRequest{
+		UserOptions: &UserOptions{
+			Email: &email,
+		},
+	}
+	err = c.put(ctx, "referral_customers/"+userId, req, &out)
+	return
+}
+
+// AddReferralCustomerCreditCard adds a credit card to a referral customer's account.
+func (c *Client) AddReferralCustomerCreditCard(referralCustomerApiKey string, creditCardOptions *CreditCardOptions, priority PaymentMethodPriority) (out *PaymentMethodObject, err error) {
+	return c.AddReferralCustomerCreditCardWithContext(context.Background(), referralCustomerApiKey, creditCardOptions, priority)
+}
+
+// AddReferralCustomerCreditCardWithContext performs the same operation as AddReferralCustomerCreditCard, but allows
+// specifying a context that can interrupt the request.
+func (c *Client) AddReferralCustomerCreditCardWithContext(ctx context.Context, referralCustomerApiKey string, creditCardOptions *CreditCardOptions, priority PaymentMethodPriority) (out *PaymentMethodObject, err error) {
+	stripeApiKeyResponse, err := c.retrieveEasypostStripeApiKey(ctx)
+	if err != nil || stripeApiKeyResponse == nil || stripeApiKeyResponse.PublicKey == "" {
+		return nil, errors.New("could not create Stripe token, please try again later")
+	}
+
+	stripeTokenResponse, err := c.createStripeToken(ctx, stripeApiKeyResponse.PublicKey, creditCardOptions)
+	if err != nil || stripeTokenResponse == nil || stripeTokenResponse.Id == "" {
+		return nil, errors.New("could not send card details to Stripe, please try again later")
+	}
+
+	return c.createEasypostCreditCard(ctx, referralCustomerApiKey, stripeTokenResponse.Id, priority)
+}
+
+func (c *Client) retrieveEasypostStripeApiKey(ctx context.Context) (out *stripeApiKeyResponse, err error) {
+	err = c.get(ctx, "partners/stripe_public_key", &out)
+	return
+}
+
+func (c *Client) createStripeToken(ctx context.Context, stripeApiKey string, creditCardOptions *CreditCardOptions) (out *stripeTokenResponse, err error) {
+	client := http.Client{}
+
+	data := url.Values{}
+	data.Set("card[number]", creditCardOptions.Number)
+	data.Set("card[exp_month]", creditCardOptions.ExpMonth)
+	data.Set("card[exp_year]", creditCardOptions.ExpYear)
+	data.Set("card[cvc]", creditCardOptions.Cvc)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://api.stripe.com/v1/tokens", strings.NewReader(data.Encode()))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Add("Authorization", "Bearer "+stripeApiKey)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+			log.Println(err) // we should never get here
```

Comment:
Issue: We don't use the `log` package in this project, let's not introduce it here. Check out other places where we raise errors.

Reply:
Resolved